# Capstone Project – The Battle of Neighborhoods | Finding a Better Place in Scarborough, Toronto

### 1. Installing and Importing Python Libraries and Dependencies

In [65]:
!pip install geocoder
!pip install folium

Importing Libraries

In [66]:
import pandas as pd
import requests
import numpy as np
import geocoder
import folium
import requests 
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import xml
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

from pandas.io.json import json_normalize 
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim 
from bs4 import BeautifulSoup

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

print("All Required Libraries Imported!")

All Required Libraries Imported!


### 2.  Data Extraction and Cleaning

Using BeautifulSoup Scraping List of Postal Codes of Given Wikipedia Page. 
Link: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [67]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
extracting_data = requests.get(url).text
wiki_data = BeautifulSoup(extracting_data, 'lxml')

Converting content of PostalCode HTML table as dataframe

In [68]:
column_names = ['Postalcode','Borough','Neighborhood']
toronto = pd.DataFrame(columns = column_names)

content = wiki_data.find('div', class_='mw-parser-output')
table = content.table.tbody
postcode = 0
borough = 0
neighborhood = 0

for tr in table.find_all('tr'):
    i = 0
    for td in tr.find_all('td'):
        if i == 0:
            postcode = td.text
            i = i + 1
        elif i == 1:
            borough = td.text
            i = i + 1
        elif i == 2: 
            neighborhood = td.text.strip('\n').replace(']','')
    toronto = toronto.append({'Postalcode': postcode,'Borough': borough,'Neighborhood': neighborhood},ignore_index=True)

In [69]:
# clean dataframe 
toronto = toronto[toronto.Borough!='Not assigned']
toronto = toronto[toronto.Borough!= 0]
toronto.reset_index(drop = True, inplace = True)
i = 0
for i in range(0,toronto.shape[0]):
    if toronto.iloc[i][2] == 'Not assigned':
        toronto.iloc[i][2] = toronto.iloc[i][1]
        i = i+1

In [70]:
df = toronto.groupby(['Postalcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df.head()

Postalcode         Borough                              Neighborhood
0      M1A\n  Not assigned\n                                          
1      M1B\n   Scarborough\n                           Malvern / Rouge
2      M1C\n   Scarborough\n  Rouge Hill / Port Union / Highland Creek
3      M1E\n   Scarborough\n       Guildwood / Morningside / West Hill
4      M1G\n   Scarborough\n                                    Woburn

In [71]:
df.describe()

Postalcode         Borough Neighborhood
count         180             180          180
unique        180              11           99
top         M1R\n  Not assigned\n             
freq            1              77           77

In [72]:
df = df.dropna()
empty = 'Not assigned'
df = df[(df.Postalcode != empty ) & (df.Borough != empty) & (df.Neighborhood != empty)]

In [73]:
df.head()

Postalcode         Borough                              Neighborhood
0      M1A\n  Not assigned\n                                          
1      M1B\n   Scarborough\n                           Malvern / Rouge
2      M1C\n   Scarborough\n  Rouge Hill / Port Union / Highland Creek
3      M1E\n   Scarborough\n       Guildwood / Morningside / West Hill
4      M1G\n   Scarborough\n                                    Woburn

In [74]:
def neighborhood_list(grouped):    
    return ', '.join(sorted(grouped['Neighborhood'].tolist()))
                    
grp = df.groupby(['Postalcode', 'Borough'])
df_2 = grp.apply(neighborhood_list).reset_index(name='Neighborhood')

In [75]:
df_2.describe()

Postalcode         Borough Neighborhood
count         180             180          180
unique        180              11           99
top         M1R\n  Not assigned\n             
freq            1              77           77

In [76]:
print(df_2.shape)
df_2.head()

(180, 3)


Postalcode         Borough                              Neighborhood
0      M1A\n  Not assigned\n                                          
1      M1B\n   Scarborough\n                           Malvern / Rouge
2      M1C\n   Scarborough\n  Rouge Hill / Port Union / Highland Creek
3      M1E\n   Scarborough\n       Guildwood / Morningside / West Hill
4      M1G\n   Scarborough\n                                    Woburn

In [77]:
def get_latilong(postal_code):
    lati_long_coords = None
    while(lati_long_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(postal_code))
        lati_long_coords = g.latlng
    return lati_long_coords
    
get_latilong('M4G')

[43.70941386000004, -79.36309957799995]

In [78]:
# Retrieving Postal Code Co-ordinates
postal_codes = df_2['Postalcode']    
coords = [ get_latilong(postal_code) for postal_code in postal_codes.tolist() ]

In [79]:
# Adding Columns Latitude & Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])
df_2['Latitude'] = df_coords['Latitude']
df_2['Longitude'] = df_coords['Longitude']

In [80]:
df_2[df_2.Postalcode == 'M5G']

Empty DataFrame
Columns: [Postalcode, Borough, Neighborhood, Latitude, Longitude]
Index: []

In [81]:
df_2.head(10)

Postalcode         Borough  \
0      M1A\n  Not assigned\n   
1      M1B\n   Scarborough\n   
2      M1C\n   Scarborough\n   
3      M1E\n   Scarborough\n   
4      M1G\n   Scarborough\n   
5      M1H\n   Scarborough\n   
6      M1J\n   Scarborough\n   
7      M1K\n   Scarborough\n   
8      M1L\n   Scarborough\n   
9      M1M\n   Scarborough\n   

                                        Neighborhood   Latitude  Longitude  
0                                                     43.648690 -79.385440  
1                                    Malvern / Rouge  43.808626 -79.189913  
2           Rouge Hill / Port Union / Highland Creek  43.785779 -79.157368  
3                Guildwood / Morningside / West Hill  43.765806 -79.185284  
4                                             Woburn  43.771545 -79.218135  
5                                          Cedarbrae  43.768791 -79.238813  
6                                Scarborough Village  43.744203 -79.228725  
7      Kennedy Park / Ionview / East Birchmount Park  43.726881 -79.265694  
8                  Golden Mile / Clairlea / Oakridge  43.713340 -79.284942  
9  Cliffside / Cliffcrest / Scarborough Village West  43.723538 -79.228353

In [82]:
address = 'Scarborough,Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude_x = location.latitude
longitude_y = location.longitude
print('The Geograpical Co-ordinate of Seattle,Washington are {}, {}.'.format(latitude_x, longitude_y))

The Geograpical Co-ordinate of Seattle,Washington are 43.773077, -79.257774.


### 3. Map of Scarborough

In [83]:
map_Scarborough = folium.Map(location=[latitude_x, longitude_y], zoom_start=8)

for lat, lng, nei in zip(df_2['Latitude'], df_2['Longitude'], df_2['Neighborhood']):
    
    label = '{}'.format(nei)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Scarborough)  
    
map_Scarborough

In [84]:
address = 'Scarborough,Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude_n1 = location.latitude
longitude_n1 = location.longitude
print('The Geograpical Co-ordinate of Neighborhood_1 are {}, {}.'.format(latitude_x, longitude_y))

The Geograpical Co-ordinate of Neighborhood_1 are 43.773077, -79.257774.


In [85]:
# @hiddel_cell
CLIENT_ID = 'ZVRXMBIPGZFKIKQ3MUV1HONJN2PHBBBG0CB3C151XWSEUQH4' # my Foursquare ID
CLIENT_SECRET = 'S1Y0JM3M5PLY4I4QABAALNDPRINPA3HBMIIHM1N31OQT3CHQ' # my Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: '+CLIENT_ID)
print('CLIENT_SECRET: '+CLIENT_SECRET)

Your credentails:
CLIENT_ID: ZVRXMBIPGZFKIKQ3MUV1HONJN2PHBBBG0CB3C151XWSEUQH4
CLIENT_SECRET: S1Y0JM3M5PLY4I4QABAALNDPRINPA3HBMIIHM1N31OQT3CHQ


In [86]:
radius = 700 
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude_n1, 
   longitude_n1, 
    radius, 
   LIMIT)
results = requests.get(url).json()

In [87]:
venues=results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues)
nearby_venues.columns

Index(['reasons.count', 'reasons.items', 'referralId', 'venue.categories',
       'venue.events.count', 'venue.events.summary', 'venue.id',
       'venue.location.address', 'venue.location.cc', 'venue.location.city',
       'venue.location.country', 'venue.location.crossStreet',
       'venue.location.distance', 'venue.location.formattedAddress',
       'venue.location.labeledLatLngs', 'venue.location.lat',
       'venue.location.lng', 'venue.location.neighborhood',
       'venue.location.postalCode', 'venue.location.state', 'venue.name',
       'venue.photos.count', 'venue.photos.groups', 'venue.venuePage.id'],
      dtype='object')

In [88]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### 4. Nearby Venues/Locations

In [89]:
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head()

venue.name  \
0               Disney Store   
1                    SEPHORA   
2  American Eagle Outfitters   
3   St. Andrews Fish & Chips   
4     Canyon Creek Chophouse   

                                    venue.categories  venue.location.lat  \
0  [{'id': '4bf58dd8d48988d1f3941735', 'name': 'T...           43.775537   
1  [{'id': '4bf58dd8d48988d10c951735', 'name': 'C...           43.775017   
2  [{'id': '4bf58dd8d48988d103951735', 'name': 'C...           43.776012   
3  [{'id': '4edd64a0c7ddd24ca188df1a', 'name': 'F...           43.771865   
4  [{'id': '4bf58dd8d48988d1cc941735', 'name': 'S...           43.776959   

   venue.location.lng  
0          -79.256833  
1          -79.258109  
2          -79.258334  
3          -79.252645  
4          -79.261694

### 5. Categories of Nearby Venues/Locations

In [90]:
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(5)

name         categories        lat        lng
0               Disney Store   Toy / Game Store  43.775537 -79.256833
1                    SEPHORA     Cosmetics Shop  43.775017 -79.258109
2  American Eagle Outfitters     Clothing Store  43.776012 -79.258334
3   St. Andrews Fish & Chips  Fish & Chips Shop  43.771865 -79.252645
4     Canyon Creek Chophouse         Steakhouse  43.776959 -79.261694

In [91]:
# Top 10 Categories
a=pd.Series(nearby_venues.categories)
a.value_counts()[:10]

Clothing Store        7
Coffee Shop           5
Restaurant            4
Pharmacy              2
Sandwich Place        2
Gas Station           2
Cosmetics Shop        2
Tea Room              2
Juice Bar             1
Mexican Restaurant    1
Name: categories, dtype: int64

In [92]:
def getNearbyVenues(names, latitudes, longitudes, radius=700):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # making GET request
        venue_results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in venue_results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [93]:
# Nearby Venues
Scarborough_venues = getNearbyVenues(names=df_2['Neighborhood'],
                                   latitudes=df_2['Latitude'],
                                   longitudes=df_2['Longitude']
                                  )


Malvern / Rouge
Rouge Hill / Port Union / Highland Creek
Guildwood / Morningside / West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park / Ionview / East Birchmount Park
Golden Mile / Clairlea / Oakridge
Cliffside / Cliffcrest / Scarborough Village West
Birch Cliff / Cliffside West
Dorset Park / Wexford Heights / Scarborough Town Centre
Wexford / Maryvale
Agincourt
Clarks Corners / Tam O'Shanter / Sullivan
Milliken / Agincourt North / Steeles East / L'Amoreaux East
Steeles West / L'Amoreaux West
Upper Rouge







Hillcrest Village
Fairview / Henry Farm / Oriole
Bayview Village
York Mills / Silver Hills
Willowdale / Newtonbrook
Willowdale
York Mills West
Willowdale







Parkwoods
Don Mills
Don Mills


Bathurst Manor / Wilson Heights / Downsview North
Northwood Park / York University
Downsview
Downsview
Downsview
Downsview









Victoria Village
Parkview Hill / Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West / Riverdale
I

In [94]:
print('There are {} Uniques Categories.'.format(len(Scarborough_venues['Venue Category'].unique())))
Scarborough_venues.groupby('Neighborhood').count().head()

There are 314 Uniques Categories.


Neighborhood Latitude  \
Neighborhood                                                               
                                                                    7700   
Agincourt                                                             26   
Alderwood / Long Branch                                               12   
Bathurst Manor / Wilson Heights / Downsview North                     24   
Bayview Village                                                        5   

                                                   Neighborhood Longitude  \
Neighborhood                                                                
                                                                     7700   
Agincourt                                                              26   
Alderwood / Long Branch                                                12   
Bathurst Manor / Wilson Heights / Downsview North                      24   
Bayview Village                                                         5   

                                                   Venue  Venue Latitude  \
Neighborhood                                                               
                                                    7700            7700   
Agincourt                                             26              26   
Alderwood / Long Branch                               12              12   
Bathurst Manor / Wilson Heights / Downsview North     24              24   
Bayview Village                                        5               5   

                                                   Venue Longitude  \
Neighborhood                                                         
                                                              7700   
Agincourt                                                       26   
Alderwood / Long Branch                                         12   
Bathurst Manor / Wilson Heights / Downsview North               24   
Bayview Village                                                  5   

                                                   Venue Category  
Neighborhood                                                       
                                                             7700  
Agincourt                                                      26  
Alderwood / Long Branch                                        12  
Bathurst Manor / Wilson Heights / Downsview North              24  
Bayview Village                                                 5

### One Hot Encoding of Features

In [95]:
# one hot encoding
Scarborough_onehot = pd.get_dummies(Scarborough_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Scarborough_onehot['Neighborhood'] = Scarborough_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Scarborough_onehot.columns[-1]] + list(Scarborough_onehot.columns[:-1])
Scarborough_onehot = Scarborough_onehot[fixed_columns]
Scarborough_grouped = Scarborough_onehot.groupby('Neighborhood').mean().reset_index()
Scarborough_onehot.head(5)

Yoga Studio  Accessories Store  Afghan Restaurant  African Restaurant  \
0            0                  0                  0                   0   
1            0                  0                  0                   0   
2            0                  0                  0                   0   
3            0                  0                  0                   0   
4            0                  0                  0                   0   

   Airport  American Restaurant  Arcade  Art Gallery  Art Museum  \
0        0                    0       0            0           0   
1        0                    0       0            0           0   
2        0                    0       0            0           0   
3        0                    0       0            0           0   
4        0                    0       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                    0                 0                   0                0   
1                    0                 0                   0                0   
2                    0                 0                   0                0   
3                    0                 0                   0                0   
4                    0                 0                   0                0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  Beach Bar  \
0                 0                 0                   0      0          0   
1                 0                 0                   0      0          0   
2                 0                 0                   0      0          0   
3                 0                 0                   0      0          0   
4                 0                 0                   0      0          0   

   Beer Bar  Beer Store  Belgian Restaurant  Big Box Store  \
0         0           0                   0              0   
1         0           0                   0              0   
2         0           0                   0              0   
3         0           0                   0              0   
4         0           0                   0              0   

   Bike Rental / Bike Share  Bike Shop  Bistro  Board Shop  Bookstore  \
0                         0          0       0           0          0   
1                         0          0       0           0          0   
2                         0          0       0           0          0   
3                         0          0       0           0          0   
4                         0          0       0           0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0         0                     0               0        0            0   
1         0                     0               0        0            0   
2         0                     0               0        0            0   
3         0                     0               0        0            0   
4         0                     0               0        0            0   

   Bridge  Bubble Tea Shop  Buffet  Building  Burger Joint  Burrito Place  \
0       0                0       0         0             0              0   
1       0                0       0         0             0              0   
2       0                0       0         0             0              0   
3       0                0       0         0             0              0   
4       0                0       0         0             0              0   

In [96]:
num_top_venues = 5
for hood in Scarborough_grouped['Neighborhood']:
    print("---- "+hood+" ----")
    temp =Scarborough_grouped[Scarborough_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----  ----
         venue  freq
0  Coffee Shop  0.09
1         Café  0.05
2   Restaurant  0.04
3        Hotel  0.04
4          Bar  0.03


---- Agincourt ----
                  venue  freq
0         Shopping Mall  0.08
1    Chinese Restaurant  0.08
2  Hong Kong Restaurant  0.04
3        Sandwich Place  0.04
4      Sushi Restaurant  0.04


---- Alderwood / Long Branch ----
               venue  freq
0       Skating Rink  0.08
1        Coffee Shop  0.08
2  Convenience Store  0.08
3               Pool  0.08
4                Pub  0.08


---- Bathurst Manor / Wilson Heights / Downsview North ----
               venue  freq
0               Bank  0.08
1        Coffee Shop  0.08
2  Convenience Store  0.04
3      Deli / Bodega  0.04
4        Bridal Shop  0.04


---- Bayview Village ----
                        venue  freq
0                        Park   0.6
1  Construction & Landscaping   0.2
2                       Trail   0.2
3     New American Restaurant   0.0
4                Optical Shop  

               venue  freq
0              Field  0.14
1              Trail  0.14
2  Convenience Store  0.14
3       Hockey Arena  0.14
4      Grocery Store  0.14


---- India Bazaar / The Beaches West ----
                  venue  freq
0     Indian Restaurant  0.06
1           Coffee Shop  0.06
2                   Gym  0.04
3             Pet Store  0.04
4  Fast Food Restaurant  0.04


---- Islington Avenue ----
           venue  freq
0       Pharmacy  0.25
1  Shopping Mall  0.12
2  Grocery Store  0.12
3           Bank  0.12
4           Park  0.12


---- Kennedy Park / Ionview / East Birchmount Park ----
                venue  freq
0      Discount Store  0.14
1  Chinese Restaurant  0.14
2         Coffee Shop  0.14
3   Convenience Store  0.07
4         Bus Station  0.07


---- Kensington Market / Chinatown / Grange Park ----
                           venue  freq
0                           Café  0.07
1  Vegetarian / Vegan Restaurant  0.07
2                    Coffee Shop  0.04
3        

            venue  freq
0             Pub  0.17
1  Clothing Store  0.08
2          Bakery  0.08
3             Spa  0.08
4     Flower Shop  0.08


---- The Danforth West / Riverdale ----
              venue  freq
0  Greek Restaurant  0.18
1              Café  0.09
2        Restaurant  0.06
3    Ice Cream Shop  0.06
4              Bank  0.06


---- The Kingsway / Montgomery Road  / Old Mill North ----
                venue  freq
0              Bakery  0.07
1      Breakfast Spot  0.07
2  Italian Restaurant  0.07
3                Bank  0.07
4    Sushi Restaurant  0.07


---- Thorncliffe Park ----
               venue  freq
0  Indian Restaurant  0.10
1               Park  0.10
2      Grocery Store  0.10
3                Gym  0.05
4             Bridge  0.05


---- Toronto Dominion Centre / Design Exchange ----
                venue  freq
0         Coffee Shop  0.10
1               Hotel  0.08
2                Café  0.07
3          Restaurant  0.05
4  Seafood Restaurant  0.03


---- Universit

In [97]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Most Common venues near neighborhood

In [98]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Scarborough_grouped['Neighborhood']

for ind in np.arange(Scarborough_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Scarborough_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                                              Coffee Shop   
1                                          Agincourt    Chinese Restaurant   
2                            Alderwood / Long Branch                  Pool   
3  Bathurst Manor / Wilson Heights / Downsview North           Coffee Shop   
4                                    Bayview Village                  Park   

        2nd Most Common Venue 3rd Most Common Venue     4th Most Common Venue  \
0                        Café                 Hotel                Restaurant   
1               Shopping Mall  Hong Kong Restaurant  Mediterranean Restaurant   
2                         Pub          Dance Studio               Coffee Shop   
3                        Bank           Bridal Shop            Sandwich Place   
4  Construction & Landscaping                 Trail             Women's Store   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0                   Bar                   Beer Bar      Asian Restaurant   
1           Supermarket  Latin American Restaurant        Breakfast Spot   
2        Sandwich Place                        Gym          Skating Rink   
3      Sushi Restaurant                Supermarket   Fried Chicken Joint   
4  Ethiopian Restaurant           Doner Restaurant            Donut Shop   

  8th Most Common Venue        9th Most Common Venue  \
0                   Pub                     Tea Room   
1                Lounge                     Pharmacy   
2    Athletics & Sports                     Pharmacy   
3         Shopping Mall                  Men's Store   
4   Dumpling Restaurant  Eastern European Restaurant   

      10th Most Common Venue  
0                     Bakery  
1               Skating Rink  
2          Convenience Store  
3  Middle Eastern Restaurant  
4          Electronics Store

### K-Means Clustering Approach

In [99]:
# Using K-Means to cluster neighborhood into 3 clusters
Scarborough_grouped_clustering = Scarborough_grouped.drop('Neighborhood', 1)
kmeans = KMeans(n_clusters=3, random_state=0).fit(Scarborough_grouped_clustering)
kmeans.labels_

array([2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 0, 1, 0, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 0, 2, 1, 2, 2, 2, 2, 2, 2, 0, 2, 2,
       2, 0, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 0, 0, 2, 2, 0], dtype=int32)

In [100]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Scarborough_merged =df_2.iloc[:16,:]

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Scarborough_merged = Scarborough_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Scarborough_merged.head()# check the last columns!

Postalcode         Borough                              Neighborhood  \
0      M1A\n  Not assigned\n                                             
1      M1B\n   Scarborough\n                           Malvern / Rouge   
2      M1C\n   Scarborough\n  Rouge Hill / Port Union / Highland Creek   
3      M1E\n   Scarborough\n       Guildwood / Morningside / West Hill   
4      M1G\n   Scarborough\n                                    Woburn   

    Latitude  Longitude  Cluster Labels       1st Most Common Venue  \
0  43.648690 -79.385440               2                 Coffee Shop   
1  43.808626 -79.189913               1  Construction & Landscaping   
2  43.785779 -79.157368               0                         Bar   
3  43.765806 -79.185284               2                 Pizza Place   
4  43.771545 -79.218135               0                 Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                 Hotel            Restaurant   
1                  Park                 Trail         Women's Store   
2                  Park         Women's Store  Ethiopian Restaurant   
3                  Park  Fast Food Restaurant                  Bank   
4                  Park      Business Service         Women's Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar              Beer Bar      Asian Restaurant   
1  Ethiopian Restaurant      Doner Restaurant            Donut Shop   
2      Doner Restaurant            Donut Shop   Dumpling Restaurant   
3           Coffee Shop            Restaurant        Breakfast Spot   
4  Ethiopian Restaurant            Donut Shop   Dumpling Restaurant   

         8th Most Common Venue        9th Most Common Venue  \
0                          Pub                     Tea Room   
1          Dumpling Restaurant  Eastern European Restaurant   
2  Eastern European Restaurant            Electronics Store   
3                   Beer Store                    Juice Bar   
4  Eastern European Restaurant            Electronics Store   

  10th Most Common Venue  
0                 Bakery  
1      Electronics Store  
2      Elementary School  
3    Fried Chicken Joint  
4      Elementary School

### Map of Clusters

In [101]:
kclusters = 10

In [102]:
# create map
map_clusters = folium.Map(location=[latitude_x, longitude_y], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
colors_array = cm.rainbow(np.linspace(0, 1, kclusters))
rainbow = [colors.rgb2hex(i) for i in colors_array]
print(rainbow)
# add markers to the map

markers_colors = []
for lat, lon, nei , cluster in zip(Scarborough_merged['Latitude'], 
                                   Scarborough_merged['Longitude'], 
                                   Scarborough_merged['Neighborhood'], 
                                   Scarborough_merged['Cluster Labels']):
    label = folium.Popup(str(nei) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

['#8000ff', '#4856fb', '#10a2f0', '#2adddd', '#62fbc4', '#9cfba4', '#d4dd80', '#ffa256', '#ff562c', '#ff0000']


In [103]:
df1=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 0,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]
df2=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 1,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]
df3=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 2,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]

In [104]:
Scarborough_Avg_HousingPrice=pd.DataFrame({"Neighborhood":df_2["Neighborhood"],
                                      "Average_Housing_Price":[335000.0,286600.0,175000.0,225900.0,219400.0,
                                                               573900.0,225000.0,370500.0,370500.0,433500.0,279200.0,
                                                               279200.0,225000.0,370500.,255400.0,433500.0,433500.0,
                                                               435000.0,289500.0,265000.0,285900.0,239400.0,
                                                               589900.0,295000.0,380500.0,378500.0,438500.0,229200.0,
                                                               229200.0,365000.0,388500.,285400.0,493500.0,477500.0,378000.0,316600.0,195000.0,225900.0,219400.0,
                                                               573900.0,367000.0,370500.0,370500.0,363500.0,279200.0,
                                                               279200.0,271000.0,370500.,255400.0,383500.0,433500.0,335000.0,286600.0,185000.0,225900.0,219400.0,
                                                               573900.0,329000.0,370500.0,370500.0,533500.0,279200.0,
                                                               279200.0,375000.0,370500.,255400.0,493500.0,433500.0,335000.0,286600.0,165000.0,225900.0,219400.0,
                                                               573900.0,425000.0,370500.0,370500.0,433500.0,279200.0,
                                                               279200.0,195000.0,370500.,255400.0,403500.0,433500.0,335000.0,286600.0,187000.0,225900.0,219400.0,
                                                               573900.0,325000.0,370500.0,370500.0,333500.0,279200.0,
                                                               279200.0,289000.0,370500.,255400.0,413500.0,433500.0,254800.0
                                                              
                                                              
                                                              ]})

ValueError: array length 103 does not match index length 180

In [105]:
Scarborough_Avg_HousingPrice.set_index('Neighborhood',inplace=True,drop=True)

NameError: name 'Scarborough_Avg_HousingPrice' is not defined

In [106]:
Scarborough_Avg_HousingPrice.plot(kind='bar',figsize=(24,18),alpha=0.75)

NameError: name 'Scarborough_Avg_HousingPrice' is not defined

### School Rating by Clusters

In [107]:
clusters=pd.DataFrame({"Cluster1":df1["Neighborhood"],
                      "Cluster2":df2["Neighborhood"],
                      "Cluster4":df3["Neighborhood"]})
clusters = clusters.replace(np.nan, '', regex=True)

In [108]:
new_Scarborough=Scarborough_merged.set_index("Neighborhood",drop=True)
#Source:https://www.greatschools.org
Scarborough_school_ratings=pd.DataFrame({"Neighborhood":df["Neighborhood"],
                                      "Top School Rating":[7,9,5,8,10,10,7,10,1,2,1,2,7,2,3,2,6,
                                                           5,4,8,9,9,6,6,4,5,4,6,8,10,8,9,6,2,
                                                           10,2,5,8,9,6,6,10,8,9,1,2,3,4,5,6,9,
                                                           8,5,9,6,9,6,4,8,10,2,5,6,3,9,8,7,
                                                           7,8,5,8,9,1,5,4,7,2,3,6,6,9,4,8,7,
                                                           4,8,9,2,6,4,7,5,10,4,6,8,9,7,5,6,5,8,7
                                                           ]})

ValueError: array length 103 does not match index length 180

In [109]:
Scarborough_school_ratings.set_index('Neighborhood',inplace=True,drop=True)

NameError: name 'Scarborough_school_ratings' is not defined

In [110]:
Scarborough_school_ratings.plot(kind='bar',figsize=(16,10),color='green',alpha=0.75);

NameError: name 'Scarborough_school_ratings' is not defined

Conclusion: In this project, using k-means cluster algorithm I separated the neighborhood into 10(Ten) different clusters and for 103 different lattitude and logitude from dataset, which have very-similar neighborhoods around them. Using the charts above results presented to a particular neighborhood based on average house prices and school rating have been made.

#### Thank You! Hope My Work Was Good, I'm Still Learning!
